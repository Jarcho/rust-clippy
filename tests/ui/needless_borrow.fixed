// run-rustfix

#![allow(clippy::needless_borrowed_reference)]

fn x(y: &i32) -> i32 {
    *y
}

#[warn(clippy::all, clippy::needless_borrow)]
#[allow(unused_variables)]
fn main() {
    let a = 5;
    let b = x(&a);
    let c = x(&a);
    let s = &String::from("hi");
    let s_ident = f(&s); // should not error, because `&String` implements Copy, but `String` does not
    let g_val = g(&Vec::new()); // should not error, because `&Vec<T>` derefs to `&[T]`
    let vec = Vec::new();
    let vec_val = g(&vec); // should not error, because `&Vec<T>` derefs to `&[T]`
    h(&"foo"); // should not error, because the `&&str` is required, due to `&Trait`
    if let Some(cake) = Some(&5) {}
    let ref_a = &a;
    let garbl = match 42 {
        44 => &a,
        45 => {
            println!("foo");
            &a
        },
        46 => &a,
        #[allow(clippy::never_loop)]
        47 => loop {
            break ref_a;
        },
        48 =>
        {
            #[allow(clippy::never_loop)]
            loop {
                break ref_a;
            }
        },
        _ => panic!(),
    };
    x(&Box::new(0)); // shouldn't error.
}

fn f<T: Copy>(y: &T) -> T {
    *y
}

fn g(y: &[u8]) -> u8 {
    y[0]
}

trait Trait {}

impl<'a> Trait for &'a str {}

fn h(_: &dyn Trait) {}
#[warn(clippy::needless_borrow)]
#[allow(dead_code)]
fn issue_1432() {
    let mut v = Vec::<String>::new();
    let _ = v.iter_mut().filter(|&ref a| a.is_empty());
    let _ = v.iter().filter(|&a| a.is_empty());

    let _ = v.iter().filter(|&a| a.is_empty());
}

#[allow(dead_code)]
#[warn(clippy::needless_borrow)]
#[derive(Debug)]
enum Foo<'a> {
    Str(&'a str),
}
