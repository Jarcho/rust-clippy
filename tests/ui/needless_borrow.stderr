error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:17:15
   |
LL |     let _ = x(&&a); // warn
   |               ^^^ help: change this to: `&a`
   |
   = note: `-D clippy::needless-borrow` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::needless_borrow)]`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:23:13
   |
LL |     mut_ref(&mut &mut b); // warn
   |             ^^^^^^^^^^^ help: change this to: `&mut b`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:35:13
   |
LL |             &&a
   |             ^^^ help: change this to: `&a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:37:15
   |
LL |         46 => &&a,
   |               ^^^ help: change this to: `&a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:43:27
   |
LL |                     break &ref_a;
   |                           ^^^^^^ help: change this to: `ref_a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:50:15
   |
LL |     let _ = x(&&&a);
   |               ^^^^ help: change this to: `&a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:51:15
   |
LL |     let _ = x(&mut &&a);
   |               ^^^^^^^^ help: change this to: `&a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:52:15
   |
LL |     let _ = x(&&&mut b);
   |               ^^^^^^^^ help: change this to: `&mut b`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:53:15
   |
LL |     let _ = x(&&ref_a);
   |               ^^^^^^^ help: change this to: `ref_a`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:56:11
   |
LL |         x(&b);
   |           ^^ help: change this to: `b`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:63:13
   |
LL |     mut_ref(&mut x);
   |             ^^^^^^ help: change this to: `x`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:64:13
   |
LL |     mut_ref(&mut &mut x);
   |             ^^^^^^^^^^^ help: change this to: `x`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:65:23
   |
LL |     let y: &mut i32 = &mut x;
   |                       ^^^^^^ help: change this to: `x`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:66:23
   |
LL |     let y: &mut i32 = &mut &mut x;
   |                       ^^^^^^^^^^^ help: change this to: `x`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:75:14
   |
LL |         0 => &mut x,
   |              ^^^^^^ help: change this to: `x`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:81:14
   |
LL |         0 => &mut x,
   |              ^^^^^^ help: change this to: `x`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:93:13
   |
LL |     let _ = (&x).0;
   |             ^^^^ help: change this to: `x`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:95:22
   |
LL |     let _ = unsafe { (&*x).0 };
   |                      ^^^^^ help: change this to: `(*x)`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:105:5
   |
LL |     (&&()).foo();
   |     ^^^^^^ help: change this to: `(&())`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:114:5
   |
LL |     (&&5).foo();
   |     ^^^^^ help: change this to: `(&5)`

error: this expression creates a reference which is immediately dereferenced by the compiler
  --> $DIR/needless_borrow.rs:140:23
   |
LL |     let x: (&str,) = (&"",);
   |                       ^^^ help: change this to: `""`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:182:13
   |
LL |             (&self.f)()
   |             ^^^^^^^^^ help: change this to: `(self.f)`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:191:13
   |
LL |             (&mut self.f)()
   |             ^^^^^^^^^^^^^ help: change this to: `(self.f)`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:228:22
   |
LL |         let _ = &mut (&mut { x.u }).x;
   |                      ^^^^^^^^^^^^^^ help: change this to: `{ x.u }`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:235:22
   |
LL |         let _ = &mut (&mut { x.u }).x;
   |                      ^^^^^^^^^^^^^^ help: change this to: `{ x.u }`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:239:22
   |
LL |         let _ = &mut (&mut x.u).x;
   |                      ^^^^^^^^^^ help: change this to: `x.u`

error: this expression borrows a value the compiler would automatically borrow
  --> $DIR/needless_borrow.rs:240:22
   |
LL |         let _ = &mut (&mut { x.u }).x;
   |                      ^^^^^^^^^^^^^^ help: change this to: `{ x.u }`

error: aborting due to 27 previous errors

